misc/mdns/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+use futures::{prelude::*, task};
+use libp2p_core::{Multiaddr, PeerId};
+use multiaddr::Protocol;
+use std::{fmt, io, net::SocketAddr, str, time::Duration, time::Instant};
+use tokio_reactor::Handle;
+use tokio_timer::Interval;
+use tokio_udp::UdpSocket;
+
+pub use dns::MdnsResponseError;
+
+mod dns;
+
+/// Hardcoded name of the mDNS service. Part of the mDNS libp2p specifications.
+const SERVICE_NAME: &'static [u8] = b"_p2p._udp.local";
+/// Hardcoded name of the service used for DNS-SD.
+const META_QUERY_SERVICE: &'static [u8] = b"_services._dns-sd._udp.local";
+
+/// A running service that discovers libp2p peers and responds to other libp2p peers' queries on
+/// the local network.
+///
+/// See the crate root documentation for more info.
+pub struct MdnsService {
+    socket: UdpSocket,
+    /// Interval for sending queries.
+    query_interval: Interval,
+    /// Whether we send queries on the network at all.
+    /// Note that we still need to have an interval for querying, as we need to wake up the socket
+    /// regularly to recover from errors. Otherwise we could simply use an `Option<Interval>`.
+    silent: bool,
+    /// Buffer used for receiving data.
+    recv_buffer: [u8; 2048],
+    /// Buffers pending to send on the socket.
+    send_buffers: Vec<Vec<u8>>,
+}
+
+impl MdnsService {
+    /// Starts a new mDNS service.
+    #[inline]
+    pub fn new() -> io::Result<MdnsService> {
+        Self::new_inner(false)
+    }
+
+    /// Same as `new`, but we don't send automatically send queries on the network.
+    #[inline]
+    pub fn silent() -> io::Result<MdnsService> {
+        Self::new_inner(true)
+    }
+
+    /// Starts a new mDNS service.
+    fn new_inner(silent: bool) -> io::Result<MdnsService> {
+        let socket = {
+            #[cfg(unix)]
+            fn platform_specific(s: &net2::UdpBuilder) -> io::Result<()> {
+                net2::unix::UnixUdpBuilderExt::reuse_port(s, true)?;
+                Ok(())
+            }
+            #[cfg(not(unix))]
+            fn platform_specific(_: &net2::UdpBuilder) -> io::Result<()> { Ok(()) }
+            let builder = net2::UdpBuilder::new_v4()?;
+            builder.reuse_address(true)?;
+            platform_specific(&builder)?;
+            builder.bind(("0.0.0.0", 5353))?
+        };
+
+        let socket = UdpSocket::from_std(socket, &Handle::default())?;
+        socket.set_multicast_loop_v4(true)?;
+        socket.set_multicast_ttl_v4(255)?;
+        // TODO: correct interfaces?
+        socket.join_multicast_v4(&From::from([224, 0, 0, 251]), &From::from([0, 0, 0, 0]))?;
+
+        Ok(MdnsService {
+            socket,
+            query_interval: Interval::new(Instant::now(), Duration::from_secs(20)),
+            silent,
+            recv_buffer: [0; 2048],
+            send_buffers: Vec::new(),
+        })
+    }
+
+    /// Polls the service for packets.
+    pub fn poll(&mut self) -> Async<MdnsPacket> {
+        // Send a query every time `query_interval` fires.
```

Comment:
According to RFC 6762, exponential delays should be used for continuous querying and queries should include known answers to not cause unnecessary re-transmissions and elevated network traffic.